Align manual_mqtt alarm control panel with manual

bonzini opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Manual MQTT Alarm Control Panel

When going from state "disarmed" to any other (armed) state such as "armed_away", the state will be "arming" instead of "pending" during the transition time as set in the configuration.

When going from an armed state (such as "armed_away") to the "triggered" state the state will still be "pending" during the transition time as set in the configuration (as it was before).

State attribute "pre_pending_state" changed to "previous_state"

State attribute "post_pending_state" changed to "next_state"

config option "pending_time" is renamed to "arming_time", functionality is the same

The time the alarm stays at "pending" when triggered has changed from delay_time of the previous state + arming_time (previously known as pending_time) of the triggered state to only the delay_time of the previous state

This makes the manual MQTT alarm control panel integration consistent with the manual alarm control panel.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The manual_mqtt and manual alarm control panel integrations have diverged. The intention of the two is to be as similar as possible, in fact manual_mqtt documentation leaves most of the detail to manual.

This includes:
* changing pending_time to arming_time (#32950, this is the breaking change)
* adding state restore on startup
* adding MQTT payloads to set custom bypass and vacation mode

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24998

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bonzini committed:
manual_mqtt: add more states

The manual alarm control panel supports ARMED_CUSTOM_BYPASS and ARMED_VACATION.
Bring them over to the MQTT version.

bonzini committed:
manual_mqtt: cosmetic changes

bonzini committed:
alarm_control_panel: move ATTR_PREVIOUS_STATE, ATTR_NEXT_STATE to alarm_control_panel.const

bonzini commented:
Added more changes. At this point the easiest way to review this is to diff the two components and the corresponding tests, since the only changes basically are related to MQTT.

bonzini committed:
manual_mqtt: add previous state for TRIGGERED as well

This will be useful to properly restore timers on restart.

bonzini commented:
I tried adding a test for the two lines that are not covered, but the code is completely wrong and I'd like to fix it for manual as well, in a separate pull request. (See the top four commits [here](https://github.com/bonzini/home-assistant/commits/manual-alarm-full-restore)).

bonzini commented:
@MartinHjelmare a few years ago you reviewed my `manual`/`manual-mqtt` pull request...

bonzini commented:
Given little attention on this PR, I have opened #82990 with the `manual` bugfixes. If it is accepted before this one, I will add the relevant commits here as well.

frenck commented:
Instead of renaming the title of the PR to prefix "Draft: ", please just put the PR in actual draft.

bonzini commented:
This is now ready for review. The only difference between manual and manual_mqtt is now the MQTT support itself.

```
 alarm_control_panel.py      |  122 ++++++++++++++++++++---
 test_alarm_control_panel.py |  515 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----------------
```

jbouwh commented on homeassistant/components/alarm_control_panel/const.py:
```diff
@@ -9,6 +9,9 @@
 ATTR_CHANGED_BY: Final = "changed_by"
 ATTR_CODE_ARM_REQUIRED: Final = "code_arm_required"
 
+ATTR_PREVIOUS_STATE: Final = "previous_state"
```
Let us not attributes here.

jbouwh commented on homeassistant/components/manual_mqtt/alarm_control_panel.py:
```diff
@@ -11,18 +11,25 @@
     STATE_ALARM_ARMED_HOME,
     STATE_ALARM_ARMED_NIGHT,
+    STATE_ALARM_ARMED_VACATION,
+    STATE_ALARM_ARMING,
```
If these are new states put related changes in a separate PR

jbouwh commented on homeassistant/components/manual_mqtt/alarm_control_panel.py:
```diff
@@ -406,20 +467,39 @@ def _async_validate_code(self, code, state):
-        }
+        if self.state in (STATE_ALARM_PENDING, STATE_ALARM_ARMING):
+            return {
+                ATTR_PREVIOUS_STATE: self._previous_state,
```
Is it really necessary to change these attributes and break.

bonzini replied:
Unfortunately yes, it is. The real motivation for me to do this PR is that I would like to have alarm control panel groups (reviving my old work from #11323), and the group needs some consistency between the devices that they include. While the group code I have written can deal with a control panel that doesn't have previous/next state attributes, the behavior is inferior.

This also justifies moving them to const, because they become standardized for all alarm control panels that want to provide full functionality for groups.

I have created a single large pull request because, more than new functionality I saw this work as fixing the mistakes that led "manual" and "manual MQTT" to become unsynchronized (#10697, #17270, #32950 and #55340). It _is_ really easier to review the combined changes than any split PR, but of course I can split it if you prefer.

To be honest, I did wonder if the "manual MQTT" integration should exist at all. It seems just as easy to do it using a "normal" manual control panel plus a template alarm control panel. But it seems to have a couple hundred users according to the telemetry, so I didn't propose the deprecation route.

jbouwh reviewed (changes requested):
If you want to add new functionality, please register an issue in the UI to notify users and make sure it does not break the config of existing users.
If you want to fix anything, please apply a non breaking fix in a separate PR. 
If you want to improve code quality to non related code, do this in a separate PR.

bonzini commented:
I will open separate pull requests with (to be applied in this order):
1. the rename of PRE_PENDING to POST_PENDING and PREVIOUS to NEXT (basic requirement for the alarm group)
2. the new ARMED states and the parametrization of tests
3. the other cosmetic changes to the code
4. the breaking changes to introduce the ARMING state
5. RestoreEntity support (this might move earlier depending on how the conflicts look like).

I will keep this pull request as "perennial draft" to track the missing changes.

bonzini committed:
manual_mqtt: make state attributes consistent with manual

As part of #32950, the manual control panel\s state attributes
were renamed from pre_pending_state and post_pending_state to
previous_state and next_state.  Unfortunately, this change was
not applied to manual_mqtt.

I would like to add alarm control panel groups, and the group code
needs some consistency between the devices that they include.

So, bring over the change.

bonzini committed:
manual_mqtt: parametrize test_no_pending

bonzini committed:
manual_mqtt: parametrize test_no_pending_when_code_not_req

bonzini committed:
manual_mqtt: parametrize test_with_pending

bonzini committed:
manual_mqtt: parametrize test_with_invalid_code

bonzini committed:
manual_mqtt: parametrize test_with_template_code

bonzini committed:
manual_mqtt: parametrize test_with_specific_pending

bonzini committed:
manual_mqtt: parametrize test_arm_via_command_topic

bonzini committed:
manual_mqtt: remove unnecessary async_block_till_done from tests

bonzini committed:
manual_mqtt: cosmetic changes to tests

bonzini committed:
manual_mqtt: bring over a new test from manual

bonzini committed:
manual_mqtt: use "arming" state during transition

Commit 7bfc1d284018aae90af4cb74614bfaa58cc4224a made a major change to the
manual alarm panel state machine, but it wasn't brought over to manual
MQTT.  Do it now.

bonzini committed:
manual_mqtt: order tests in the same way as manual

bonzini committed:
manual_mqtt: add state restore

jbouwh commented:
> the breaking changes to introduce the ARMING state

Try to deprecate first to avoid breaking using `repairs`. Later you can remove the work-a-rounds.
